REGRESSION (STP 161): cache-storage 100+ test failures on WPT

youennf opened the pull request:
#### 7fa5066e9d64de34b5fa1c46bf3b8fecf3957fc2
<pre>
REGRESSION (STP 161): cache-storage 100+ test failures on WPT
<a href="https://bugs.webkit.org/show_bug.cgi?id=250590">https://bugs.webkit.org/show_bug.cgi?id=250590</a>
rdar://problem/104237218

Reviewed by Chris Dumez.

This is a regression of switching to the new CacheStorage backend in ephemeral sessions.
WPT is using ephemeral sessions which is the only code path to store ResponseResourceBase::CrossThreadData in CacheStorageMemoryStore.
It is manipulated in a WorkQueue where several threads might be involved, which is not working well with AtomString that are bound by thread.
In particular, when we create a ResponseResourceBase from a CacheStorageMemoryStore record CrossThreadData, the CrossThreadData String will be change to an AtomString.
To prevent this, we make sure to isolate copy all Strings inside ResponseResourceBase::CrossThreadData::isolatedCopy() and create a new CrossThreadData from CacheStorageMemoryStore records.

We isolate copy all strings as it is safer in general, the result of ResponseResourceBase::CrossThreadData::copy() could be sent to another thread for instance.
There should be no perf impact except on CacheStorage in ephemeral sessions since ResponseResourceBase::CrossThreadData::isolatedCopy() is only used there.

We also fix ResourceResponseBase::CrossThreadData copy as it was missing statusText copy.
And we also add a bunch of WTFMove as a small improvement in ResourceResponseBase::fromCrossThreadData.

We remove CacheStorageRecord::copy and put it as a static function in CacheStorageMemoryStore as it is only used there and is very specific in that it isolate copy the ResponseResourceBase::CrossThreadData but not other fields.

To cover this patch, we copy WPT cache-matchAll.https.any.html in http/wpt to run it with an ephemeral session.

* LayoutTests/http/wpt/cache-storage/cache-matchAll.https.any-expected.txt: Added.
* LayoutTests/http/wpt/cache-storage/cache-matchAll.https.any.html: Added.
* LayoutTests/http/wpt/cache-storage/cache-matchAll.https.any.js: Added.
* Source/WebCore/platform/network/ResourceResponseBase.cpp:
(WebCore::ResourceResponseBase::CrossThreadData::copy const):
(WebCore::ResourceResponseBase::fromCrossThreadData):
* Source/WebKit/NetworkProcess/storage/CacheStorageMemoryStore.cpp:
(WebKit::copyCacheStorageRecord):
(WebKit::CacheStorageMemoryStore::readAllRecords):
(WebKit::CacheStorageMemoryStore::readRecords):
* Source/WebKit/NetworkProcess/storage/CacheStorageRecord.h:
(WebKit::CacheStorageRecord::copy const): Deleted.

Canonical link: <a href="https://commits.webkit.org/259418@main">https://commits.webkit.org/259418@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5e534ebd5897c3fa0fbc1f7d2c220ee267930be1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104862 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13942 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37757 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114131 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174322 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108770 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15072 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4866 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97189 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113156 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110622 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/11647 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94651 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/39165 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93512 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26269 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7286 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27630 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7382 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4220 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13438 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47180 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6497 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9171 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/network/ResourceResponseBase.cpp:
```diff
@@ -131,13 +132,13 @@ ResourceResponse ResourceResponseBase::fromCrossThreadData(CrossThreadData&& dat
 
     response.setURL(data.url);
-    response.setMimeType(AtomString { data.mimeType });
+    response.setMimeType(AtomString { WTFMove(data.mimeType).isolatedCopy() });
```
I admit I don't understand this change. The whole point of CrossThreadData is that it should contain strings that have already been isolated. It seems therefore odd to require the recipient of said CrossThreadData to isolate copy those strings again. I have a suspicion there must be a better fix.

cdumez replied:
Like thread A calls r.crossThreadData() and captures this in a lambda when it dispatches to thread B.
Thread B should be able to create a response from this CrossThreadData without further isolate copying, since this was already done by the crossThreadData() on thread A. So in the general case, it should not be needed but your patch adds this cost anyway.

I suspect your use case if doing something more complicated that is slightly wrong and we should be fixing that wrong thing instead of updating fromCrossThreadData(). E.g. Maybe we have code that dispatches the CrossThreadData yet again to another thread, without isolatedCopying() it?

youennf replied:
The issue is that `AtomString(a)`, `a` being isolatedCopy will change the StringImpl of `a` and it will no longer be isolatedCopy. With a WorkQueue that might dispatch tasks in different threads, we end up into issues.

I'll change ResourceResponseBase::CrossThreadData::copy() instead of fromCrossThreadData though, this seems cleaner this way.

cdumez replied:
Sorry, I don't understand your comment.

`AtomString(a)` with `a` being an isolatedCopy will indeed copy the StringImpl and thus it won't be an isolated copy anymore. And that seems fine unless you pass `a` to another thread (without isolatedCopying() it again) later on?

youennf replied:
We have a work queue.
We dispatches a first task on this queue where we create a ResourceResponse from a CrossThreadData that is stored in a record.
At that point, the record's CrossThreadData String will be changed to an AtomString.
The same work queue dispatches a second task where we remove the record, hence we destroy the CrosssThreadData. This triggers the removal of the AtomString but from a different thread it was created on.

stwrt commented on Source/WebCore/platform/network/ResourceResponseBase.cpp:
```diff
@@ -131,13 +132,13 @@ ResourceResponse ResourceResponseBase::fromCrossThreadData(CrossThreadData&& dat
     ResourceResponse response;
 
     response.setURL(data.url);
-    response.setMimeType(AtomString { data.mimeType });
```
Why do we not need to create isolated copies for all objects like `data.url` too?

youennf replied:
AtomString is modifying the given strings, hence the issue. data.url remains an URL.
Next version will hopefully be clearer.

stwrt reviewed (commented):
> This should not have a per fimpact on other code paths

This was not super clear what you were trying to say.

cdumez commented on Source/WebCore/platform/network/ResourceResponseBase.cpp:
```diff
@@ -87,14 +87,16 @@ ResourceResponseBase::ResourceResponseBase(std::optional<ResourceResponseBase::R
 ResourceResponseBase::CrossThreadData ResourceResponseBase::CrossThreadData::copy() const
```
Seems like we should be calling this isolatedCopy() then? Maybe the copy() version that is cheaper is still valuable for call sites that don't need the isolated copying?

youennf replied:
CrossThreadData has no copy constructor by design.
The copy() version is dangerous, we should just not use it.
I prefer keeping copy() since isolatedCopy() would make think we could have a copy() version which we should not.

cdumez replied:
But I think we need a copy() function that doesn't isolate copy. You new code is slower and as far as I can tell is not required in the common case. It is still unclear to me why you need to isolate copy at all. Maybe it would be clearer if I understand what the call site was doing.

youennf replied:
> But I think we need a copy() function that doesn't isolate copy.

The only use of copy() is in CacheStorageMemoryStore::readRecords where it is unsafe.

youennf replied:
> The call site is doing.

The storage is keeping a list of record in memory, which contains ResourceResponses as CrossThreadData.
The storage is queried from time to time and answers with a ResourceResponse when there is a match.

cdumez reviewed (approved):
LGTM since I believe the fix is right, after looking through the call site. That said, I would really prefer if we renamed `copy()` to `isolatedCopy()` now that its implementation actually isolate copies everything. We have a standard naming for such functions, I don't see a clear reason here to not adhere to this naming.

Feel free to land and ignore me but I am hoping you'll agree.

szewai reviewed (approved):
> It is manipulated in a WorkQueue where several threads might be involved, which has an impact on AtomString.

Can we clarify this a bit more? Something like:

> AtomString is bounded by thread. CacheStorageMemoryStore, which stores ResponseResourceBase::CrossThreadData, is bounded by WorkQueue and can be accessed from multiple threads. In CacheStorage backend, a record reading task can get a copy of stored CrossThreadData with CrossThreadData::copy(), and create a ResponseResource from the copy with fromCrossThreadData(). 
> The issue is CrossThreadData::copy() returns a shallow copy, and fromCrossThreadData() converts String into AtomString (AtomString constructor modifies StringImpl), making stored CrossThreadData object no longer safe to be accessed from a different thread. To solve this, we now make CrossThreadData::copy() returns an isolated copy.

youennf committed:
REGRESSION (STP 161): cache-storage 100+ test failures on WPT
https://bugs.webkit.org/show_bug.cgi?id=250590
rdar://problem/104237218

Reviewed by Chris Dumez.

This is a regression of switching to the new CacheStorage backend in ephemeral sessions.
WPT is using ephemeral sessions which is the only code path to store ResponseResourceBase::CrossThreadData in CacheStorageMemoryStore.
It is manipulated in a WorkQueue where several threads might be involved, which is not working well with AtomString that are bound by thread.
In particular, when we create a ResponseResourceBase from a CacheStorageMemoryStore record CrossThreadData, the CrossThreadData String will be change to an AtomString.
To prevent this, we make sure to isolate copy all Strings inside ResponseResourceBase::CrossThreadData::isolatedCopy() and create a new CrossThreadData from CacheStorageMemoryStore records.

We isolate copy all strings as it is safer in general, the result of ResponseResourceBase::CrossThreadData::copy() could be sent to another thread for instance.
There should be no perf impact except on CacheStorage in ephemeral sessions since ResponseResourceBase::CrossThreadData::isolatedCopy() is only used there.

We also fix ResourceResponseBase::CrossThreadData copy as it was missing statusText copy.
And we also add a bunch of WTFMove as a small improvement in ResourceResponseBase::fromCrossThreadData.

We remove CacheStorageRecord::copy and put it as a static function in CacheStorageMemoryStore as it is only used there and is very specific in that it isolate copy the ResponseResourceBase::CrossThreadData but not other fields.

To cover this patch, we copy WPT cache-matchAll.https.any.html in http/wpt to run it with an ephemeral session.

* LayoutTests/http/wpt/cache-storage/cache-matchAll.https.any-expected.txt: Added.
* LayoutTests/http/wpt/cache-storage/cache-matchAll.https.any.html: Added.
* LayoutTests/http/wpt/cache-storage/cache-matchAll.https.any.js: Added.
* Source/WebCore/platform/network/ResourceResponseBase.cpp:
(WebCore::ResourceResponseBase::CrossThreadData::copy const):
(WebCore::ResourceResponseBase::fromCrossThreadData):
* Source/WebKit/NetworkProcess/storage/CacheStorageMemoryStore.cpp:
(WebKit::copyCacheStorageRecord):
(WebKit::CacheStorageMemoryStore::readAllRecords):
(WebKit::CacheStorageMemoryStore::readRecords):
* Source/WebKit/NetworkProcess/storage/CacheStorageRecord.h:
(WebKit::CacheStorageRecord::copy const): Deleted.

Canonical link: https://commits.webkit.org/259418@main
